daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversionStackSafe.scala

```diff
@@ -0,0 +1,368 @@
+                case arg :: args =>
+                  loop(Down(arg, env), Cont.App2(fun, argsDone, env, args, cont))
+              }
+
+            case Cont.Case1(env, alts, cont) =>
+              val scrut = result
+              alts match {
+                case Nil =>
+                  loop(Up(target.SECase(scrut, Nil)), cont)
+                case source.SCaseAlt(pat, rhs) :: alts =>
+                  val n = pat.numArgs
+                  loop(Down(rhs, env.shift(n)), Cont.Case2(scrut, Nil, pat, env, alts, cont))
+              }
+
+            case Cont.Case2(scrut, altsDone0, pat, env, alts, cont) =>
+              val altsDone = target.SCaseAlt(pat, result) :: altsDone0
+              alts match {
+                case Nil =>
+                  loop(Up(target.SECase(scrut, altsDone.reverse)), cont)
+                case source.SCaseAlt(pat, rhs) :: alts =>
+                  val n = pat.numArgs
+                  loop(Down(rhs, env.shift(n)), Cont.Case2(scrut, altsDone, pat, env, alts, cont))
+              }
+
+            case Cont.Let1(boundsDone0, env, bounds, body, cont) =>
+              val boundsDone = result :: boundsDone0
+              val depth = boundsDone.length
+              val env1 = env.shift(depth)
+              bounds match {
+                case Nil =>
+                  loop(Down(body, env1), Cont.Let2(boundsDone, cont))
+                case bound :: bounds =>
+                  loop(Down(bound, env1), Cont.Let1(boundsDone, env, bounds, body, cont))
+              }
+
+            case Cont.Let2(boundsDone, cont) =>
+              val body = result
+              loop(Up(target.SELet(boundsDone.reverse, body)), cont)
+
+            case Cont.TryCatch1(env, handler, cont) =>
+              val body = result
+              loop(Down(handler, env.shift(1)), Cont.TryCatch2(body, cont))
+
+            case Cont.TryCatch2(body, cont) =>
+              val handler = result
+              loop(Up(target.SETryCatch(body, handler)), cont)
+
+            case Cont.ScopeExercise(cont) =>
+              val body = result
+              loop(Up(target.SEScopeExercise(body)), cont)
+
+            case Cont.LabelClosure(label, cont) =>
+              val expr = result
+              loop(Up(target.SELabelClosure(label, expr)), cont)
+
+          }
+      }
+    }
+
+    /* The (stack-safe) transformation is started here, passing the original source
+     * expression (source1), an empty environment, and the continuation to 'Finish'.
+     */
+    loop(Down(source0, Env()), Cont.Finish)
+  }
+
+  // This is the 'freeVars' implementation from the pre-stack-safe closure-conversion code.
+  // TODO: recode to avoid polynomial-complexity of 'freeVars' computation.
+
+  /** Compute the free variables in a speedy expression.
+    * The returned free variables are de bruijn indices
+    * adjusted to the stack of the caller.
+    */
+  private[this] def freeVars(expr: source.SExpr, initiallyBound: Int): Set[Int] = {
```

Comment:
this also doesn’t look stack safe afaict

Reply:
This could probably be solved in the same time as the complexity TODO above. 
We could simply compute and store the free variables in the `source.SExpr` objects directly as we are building them in the previous pass.

Reply:
You are correct. This is not stack safe!
I added a note to this effect in issue #11830